Phase3

Doinededoidoine opened the pull request:
Le code de la phase 3 : implémentation de deux heuristiques accélératrices sur l'algorithme de Kruskal, et implémentation de l'algorithme de Prim, pour des instances de graphes symétriques.

Doinededoidoine committed:
Début de l'algorithme de Prim (non-fonctionnel)

Doinededoidoine committed:
Kruskal et ses deux heuristiques sont fonctionnels. 
Prim en cours d'implémentation :
- La PriorityQueue est implémentée
- L'algorithme de Prim est dans main3.jl
- Il reste à trouver une manière d'ordonner les noeuds sans perturber 
kruskal(), puis à tester les deux algorithmes.

Doinededoidoine committed:
Correction d'un bug sur kruskal

Doinededoidoine committed:
Prim semble fonctionner
Il reste à lancer les tests

Doinededoidoine committed:
Corrigé Prim pour ajouter d'abord les noeuds de poids MINIMUM.
Etendu PriorityQueue à l'extraction de l'objet de poids minimal : 
poplast!()

Doinededoidoine committed:
Correction : 
Prim : Le poids d'un noeud est bien mis à zéro quand il est ajouté à 
l'arbre de recouvrement minimum.

En commentaire dans les algorithmes, du code permet de vérifier que les 
heuristiques sont bien implantées.

Doinededoidoine committed:
- Changé la méthode read_nodes() dans read_stsp.jl pour attribuer des coordonnées arbitrairement nulles aux noeuds d'un graphe dans le cas où elles ne sont pas explicitées.
- Indiqué dans les docstrings de Node, kruskal() et prim() que tous les 
noeuds d'un graphe en entrée doivent avoir des attributs "name" 
différents.
- Ajouté deux tests : un graphe à 9 noeuds issu de l'exemple du cours 
sur l'algorithme de Prim, et le graphe "fri26.tsp" dont les coordonnées 
des noeuds ne sont pas explicitées.
- Ajouté un fichier benchmark.jl qui teste les performances de kruskal() 
sur différents TSP symétriques. (Ne parvient pas à redéfinir la 
constante STSP_DIR...)

Doinededoidoine committed:
Derniers changements sur becnhmark.jl. Ne marche toujours pas.

Doinededoidoine committed:
- node.jl : changé 0 pour 0.0
- parent_table.jl : implémenté une nouvelle méthode 
"update_parent_and_weight!()" qui, dans l'algorithme de Prim, affecte à 
un nouveau noeud son nouveau poids et son nouveau parent le nouveau 
poids est inférieur au poids actuel.
- main3.jl : Corrigé prim() en utilisant la méthode 
update_parent_and_weight!(). Auparavant, prim() n'ajoutait pas les 
bonnes arêtes à l'arbre de recouvrement minimum (exemple : graph_9n issu 
du cours).

Doinededoidoine committed:
Petit oubli sur le denrier push :
update_min_weight --> set_min_weight
dans la méthode prim()

Doinededoidoine committed:
Correction de fautes d'orthographes

mtanneau commented on projet/phase1/node.jl:
```diff
@@ -11,22 +13,44 @@ Exemple:
 mutable struct Node{T} <: AbstractNode{T}
   name::String
   data::T
+  min_weight::Float64
```
Ajouter un attribut à `Node` vous force à modifier vos fonctions pré-existantes (constructeurs, accesseurs, etc)...

`min_weight` n'a de sens qu'à l’intérieur de l’algorithme de Prim, et n'est pas une propriété intrinsèque d'un nœud : cette quantité dépend notamment de l'arbre partiel considéré.

Votre code serait plus simple (et plus facile à maintenir) si `min_weight` était stocké dans une structure de données séparée, qui n'existe qu'au sein de l'algorithme de Prim.

mtanneau commented on projet/phase1/read_stsp.jl:
```diff
@@ -39,14 +39,17 @@ function read_nodes(header::Dict{String}{String}, filename::String)
+  dim = parse(Int, header["DIMENSION"])
 
-
+  # Si les coordonnées des noeuds du graphe ne sont pas explicitées, on les fixe arbitrairement à [0, 0].
```
Ce cas est déjà traité par la fonction `read_nodes` : le résultat est un dictionnaire vide.

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -21,16 +21,67 @@ function kruskal(graph::AbstractGraph{T}) where T
-        if root(parent_table, node1) != root(parent_table, node2)
+        # Si les deux noeuds ont des racines différentes, on ajoute l'arête à l'arbre de recouvrement minimum puis on réunit les
+        # composantes connexes en désignant l'une des deux racines comme le nouveau parent de l'autre racine.
+        if root(parent_table, node1) !== root(parent_table, node2)
```
Testez plutôt si `root1` et `root2` ont le même nom

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -21,16 +21,67 @@ function kruskal(graph::AbstractGraph{T}) where T
     end
-    # On affiche puis on renvoie l'arbre de recouvrement minimum.
-    show(min_tree)
+    # # Le code suivant permet de tester que l'union via le rang et la compression se font correctement :
```
Ne gardez pas de code commenté comme ça.

Ce genre de test doit être fait manuellement lors des tests unitaires.

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -21,16 +21,67 @@ function kruskal(graph::AbstractGraph{T}) where T
+entrée en utilisant l'algorithme de Prim. La méthode renvoie un objet de type Graph.
+ATTENTION : dans le graphe en entrée, les noeuds doivent tous avoir un attribut "name" différent.
+"""
+function prim(graph::AbstractGraph{T}, starting_node::AbstractNode) where T
```
Ajoutez une valeur par défaut pour la racine

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -21,16 +21,67 @@ function kruskal(graph::AbstractGraph{T}) where T
+function prim(graph::AbstractGraph{T}, starting_node::AbstractNode) where T
+    parent_table = init_parent_table_prim(graph)
+    # On attribue un poids nul au noeud de départ et on stocke les noeuds dans une file de priorité par poids.
+    set_min_weight!(starting_node, 0.0)
```
Ne fonctionne pas si vous avez des arêtes avec un poids négatif.

mtanneau commented on projet/phase3/queue.jl:
```diff
@@ -0,0 +1,63 @@
+abstract type AbstractQueue{T} end
+
+"""Type représentant une file avec des éléments de type T."""
+mutable struct Queue{T} <: AbstractQueue{T}
```
Vous n'utilisez pas `Queue`

mtanneau commented on projet/phase3/queue.jl:
```diff
@@ -0,0 +1,63 @@
+            lowest = item
+        end
+    end
+    idx = findall(x -> x == lowest, q.items)[1]
```
`findfirst` serait plus efficace

mtanneau commented on projet/phase3/queue.jl:
```diff
@@ -0,0 +1,63 @@
+            highest = item
+        end
+    end
+    idx = findall(x -> x == highest, q.items)[1]
```
idem: `findfirst` est plus rapide

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -61,21 +74,19 @@ function increase_rank!(parent_table::AbstractParentTable, node::AbstractNode)
-        root = root_
+    parent_ = parent(parent_table, node)
+    if node !== parent_
+        parent_ = root(parent_table, parent_)
```
Dangereux si vous travaillez sur des graphes très larges (la pile d'appel récursifs a une taille limitée).

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -61,21 +74,19 @@ function increase_rank!(parent_table::AbstractParentTable, node::AbstractNode)
-le plus ancien parent de ce noeud.
+le plus ancien parent de ce noeud. Au passage, compresse le chemin vers la racine.
 """
 function root(parent_table::AbstractParentTable, node::AbstractNode)
```
Modifie la table des parents `parent_table`, donc mérite un `!`

mtanneau reviewed (commented):
Il faut corriger votre code pour l'algorithme de Prim (incorrect si des arêtes ont un poids négatif).

Niveau vitesse : votre code est très lent car `set_parent!` a une complexité linéaire en le nombre de nœuds du graphe. Vous pouvez réduire ça à une complexité en temps constant.

mtanneau commented on projet/phase1/node.jl:
```diff
@@ -11,22 +13,44 @@ Exemple:
+
+"""Affecte une nouvelle valeur au poids d'un noeud."""
+function set_min_weight!(node::AbstractNode, weight::Float64)
+  node.min_weight = max(0.0, weight)
```
Il ne faut pas forcer `min_weight` à prendre des valeurs positives : vous pouvez avoir des arêtes de poids négatif dans le graphe.

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -21,16 +21,67 @@ function kruskal(graph::AbstractGraph{T}) where T
+ATTENTION : dans le graphe en entrée, les noeuds doivent tous avoir un attribut "name" différent.
+"""
+function prim(graph::AbstractGraph{T}, starting_node::AbstractNode) where T
+    parent_table = init_parent_table_prim(graph)
```
Assurez-vous de bien réinitialiser tous les `min_weight` à `Inf` avant de commencer.

Doinededoidoine committed:
Corrections de la phase 3 :
+ arêtes de poids négatif
- structure "Queue"
- code en commentaire dans "main3.jl"

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -80,19 +80,19 @@ end
     parent_ = parent(parent_table, node)
-    if node !== parent_
-        parent_ = root(parent_table, parent_)
+    if name(node) !== name(parent_)
```
`!=`

mtanneau commented on projet/phase3/main3.jl:
```diff
@@ -23,65 +23,51 @@ function kruskal(graph::AbstractGraph{T}) where T
-    # # On affiche puis on renvoie l'arbre de recouvrement minimum.
-    # show(min_tree)
+    # On affiche puis on renvoie l'arbre de recouvrement minimum.
+    show(min_tree)
```
Évitez d'afficher l'arbre à chaque appel, `kruskal` sera amenée à être exécutée de nombreuses fois.
